pmi: making pmi a sub module

hzhou opened the pull request:
## Pull Request Description
Making PMI a submodule will allow the use and testing of PMI functions without MPI. It will facilitate the standardization of PMI interface.

This PR is based on #5884 

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

hzhou commented:
test:mpich/ch3/tcp

hzhou commented:
test:mpich/ch4/pmi

hzhou commented:
test:mpich/ch4/pmi

hzhou commented:
Rebased --
test:mpich/ch3/most
test:mpich/ch4/most
test:mpich/pmi

raffenet commented on src/pmi/examples/test_pmi1.c:
```diff
@@ -0,0 +1,49 @@
+
+    if (rank == 0) {
+        /* NOTE: PMI-v1 does not allow space in both key and value */
+        PMI_KVS_Put(kvsname, "key0", "got-key0");
```
It's a no-op in Simple PMI, but I think technically we should have a `PMI_KVS_Commit()` here.

hzhou replied:
Yeah, I meant to discuss this --

PMI_KVS_Commit() flushed the local put messages to global KVS space, so it is useful if we want some sort of p2p kvs exchange where one process `Put-Commit` and another process `Get (wait)`. It is useful since it avoids global `PMI_Barrier`. On the other hand, if the application is to do global KVS exchange, a single PMI_Barrier is more efficient than individual processes each try to push their own message (on to other processes' KVS). That is, adding `PMI_KVS_Commit` before a `PMI_Barrier` can be costly if we do not implement it as a "no-op" (one of my plan).

Is there text somewhere guarantee a `PMI_KVS_Get` will get the message of a `PMI_KVS_Put` before a barrier? If there is, then it is legally okay to omit `PMI_KVS_Commit` here.

raffenet replied:
The Notes section for `PMI_KVS_Put` in pmi.h says

> This function puts the key/value pair in the specified keyval space.  The
value is not visible to other processes until 'PMI_KVS_Commit()' is called.

So I think we are stuck requiring it before any other process can get the value. I do think as things are defined, a barrier could be avoided if the getting process just keeps calling `PMI_KVS_Get` until the value is available.

hzhou replied:
We can treat those text as "standard" and see it as "stuck", or, since there was never an official standard, we can modify the text now :) . As long as the changes are backward compatible, I would think it'll be okay. What do you think?

hzhou replied:
The word I want to add is a note to the `PMI_Barrier`, that `PMI_Barrier` automatically applies `PMI_KVS_Commit` for each process.

raffenet replied:
I see. I think that is essentially the idea behind the semantic of `PMI2_KVS_Fence`. Then again, PMI2 doesn't have a non-collective API for making KV pairs visible.

The quoted text exists in various implementations, though I'd bet `PMI_KVS_Commit` is a no-op in all those cases (Slurm ✅  Microsoft ✅  Flux ✅  Cray ❓ ). Flux actually takes it a step further and states that KV pairs are not visible until after _both_ a commit and a barrier https://flux-framework.readthedocs.io/projects/flux-rfc/en/latest/spec_13.html#key-value-store.

The PMI server in Hydra does not make any keys available until after a barrier, so I guess it's an OK change to make 🤷 .

raffenet commented on configure.ac:
```diff
@@ -1529,55 +1615,45 @@ else
+            AC_MSG_ERROR([$pm_name requires PMI-v1, but $with_pmi was chosen.])
+        fi
+        ;;
+      remshell)
```
nitpick in the commit message i think gforker2 is supposed to be remshell
```
This commit makes 3 orthogonal options:
    * --with-pmi={pmi1, pmi2, pmix}
    * --with-pmilib={mpich, slurm, cray, pmix}
    * --with-pm={no,hydra,hydra2,gforker,gforker2}
```

hzhou replied:
Fixed.

raffenet reviewed (commented):
just a couple of minor comments

hzhou commented:
Added new notes on the implications of `PMI_Barrier` and `PMI_KVS_Commit`.

Added patches for PMI spawn to work with either pmi1, pmi2, or Slurm since they all use different info type.

test:mpich/ch3/most
test:mpich/ch4/most
test:mpich/pmi

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most
test:mpich/pmi

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most
test:mpich/pmi

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most
test:mpich/pmi

hzhou commented:
test:mpich/ch3/most
test:mpich/ch4/most
test:mpich/pmi

hzhou commented:
Removed the Slrum fix (moved to separate PR) 

test:mpich/pmi

raffenet commented on src/pmi/include/pmi.h:
```diff
@@ -235,7 +238,7 @@ Return values:
 This function is a collective call across all processes in the process group
 the local process belongs to.  It will not return until all the processes
-have called 'PMI_Barrier()'.
+have called 'PMI_Barrier()'. 'PMI_Barrier' automatically applies 'PMI_KVS_Put'.
```
```suggestion
have called 'PMI_Barrier()'. 'PMI_Barrier' automatically applies 'PMI_KVS_Commit'.
```

hzhou replied:
Good catch! Now fixed.
